### PR TITLE
WIP: Add spawn class based on subprocess.Popen

### PR DESCRIPTION
**Pull request by takluyver**
This will ultimately allow simple uses of Pexpect on Windows. Ping @blink1073 - I copied your gist for the implementation of this.

Open questions:

- At present, `__init__.py` imports `pty_spawn`, so you can't even load Pexpect on Windows. I started doing something along the lines of:

```python
if sys.platform == 'win32':
    from .popen_spawn import PopenSpawn as spawn
else:
    from .pty_spawn import spawn
```

But that didn't feel quite right; an apparently similar API would be doing something very different internally, which would doubtless lead to surprising bugs. Maybe on Windows the top-level package simply shouldn't have `spawn` and `spawnu`, and we'll leave them in place on Unix for backwards compatibility.

- I'm aiming to make it as close as possible to the applicable parts of the pty_spawn API, but in a couple of places, the Unixy nature of that API conflicts with Popen and cross platform support:
   1. `spawn.kill()` takes a signal number. `Popen()` has `.kill()` and `.terminate()`, which send fixed Unix signals (SIGKILL and SIGTERM), and both do TerminateProcess on Windows.
   2. `spawn.wait()` returns the raw exit code, a 2-byte number where the high byte is the terminating signal number. `Popen.wait()` uses negative numbers for terminating signals.

   My inclination in these cases is that PopenSpawn should stay closer to the cross platform API of Popen, rather than trying to closely copy the pty_spawn interface and being inconsistent between platforms.

### CODE REVIEW

**Comment by jquast** [comment]
I've reviewed this PR, +1

**Comment by takluyver** [comment]
PopenSpawn seems to be hitting EOF prematurely on PyPy. Not sure what's causing that at the moment.

**Comment by blink1073** [comment]
I have no experience with PyPy, does it handle `b''` strings?

**Comment by takluyver** [comment]
I think so - I'd expect a very different error message if that was the problem. Maybe it's some kind of race condition, but I can't think what.

File: pexpect/popen_spawn.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        self._log(buf, 'read')
+        return buf
+
+    def _read_incoming(self):
+        """Run in a thread to move output from a pipe to a queue."""
+        fileno = self.proc.stdout.fileno()
+        while 1:
+            buf = b''
+            try:
+                buf = os.read(fileno, 1024)
```
Comment (jquast) [inline pexpect/popen_spawn.py]: @blink1073 why did you add the ``time.sleep(0.001)`` call after queue.put? I consider your original commit also correct: as our read call is always blocking here, there is no need to artificially prevent a "busyloop thread", or is there a consideration here?

**Comment by jquast** [comment]
Re-reviewed with fine comb for cause of EOF on PyPy cause, I don't see it, unfortunately. I'd sooner add another SkipTest for PyPy.

File: pexpect/popen_spawn.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            except OSError as e:
+                self._log(e, 'read')
+
+            if not buf:
+                self._read_queue.put(None)
+                return
+
+            self._read_queue.put(buf)
+            time.sleep(0.001)
```
Comment (jquast) [inline pexpect/popen_spawn.py]: @blink1073 why did you add the ``time.sleep(0.001)`` call after queue.put? I consider your original commit also correct: as our read call is always blocking here, there is no need to artificially prevent a "busyloop thread", or is there a consideration here? [edit: cite correct line number]

Comment (blink1073) [inline pexpect/popen_spawn.py]: Just defensive programming as far as I remember.

Comment (takluyver) [inline pexpect/popen_spawn.py]: I'm going to try removing it to see if it makes a difference to the failing test - if it doesn't, I'll remove the commit.

**Comment by takluyver** [comment]
Taking the sleep out appears to have made the same failure appear on CPython. Can anyone make anything of that?

**Comment by blink1073** [comment]
Thoughts: `bufsize = 0` results in a busy-loop, and pypy "optimizes" the `sleep` out?

**Comment by takluyver** [comment]
It turned out that it was revealing a real bug. When the reader thread reached EOF, `read_nonblocking()` could raise EOF while it had buffered data that had not yet been passed on to the caller. There were also different stages of the pipeline clobbering the `.buffer` attribute for one another.

I think I have fixed these things now, and the tests are passing even on PyPy. But it's very tricky to work on this code and keep all the different moving bits in my head (it's not spaghetti code, but lasagne code - lots of little layers). I feel like a much cleaner, less stateful API could be designed to do what's needed, but I'm not sure I want to write a successor to Pexpect.

**Comment by blink1073** [comment]
Tricky, but your fix looks good to me.
